db/migrations/20240802044339_player_use_user_id_over_username.go

```diff
@@ -0,0 +1,116 @@
+package migrations
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/pressly/goose/v3"
+)
+
+func init() {
+	goose.AddMigrationContext(upPlayerUseUserIdOverUsername, downPlayerUseUserIdOverUsername)
+}
+
+func upPlayerUseUserIdOverUsername(ctx context.Context, tx *sql.Tx) error {
+	// Additional note; we also rename ip_address to IP. For some bizarre
+	// reason, ip_address is never serialized
+	_, err := tx.ExecContext(ctx, `
+create table player_dg_tmp
+(
+	id varchar(255) not null
+		primary key,
+	name varchar not null,
+	user_agent varchar,
+	user_id varchar not null
+		references user (id)
+			on update cascade on delete cascade,
+	client varchar not null,
+	ip varchar,
+	last_seen timestamp,
+	max_bit_rate int default 0,
+	transcoding_id varchar,
+	report_real_path bool default FALSE not null,
+	scrobble_enabled bool default true
+);
+
+insert into player_dg_tmp(id, name, user_agent, user_id, client, ip, last_seen, max_bit_rate, transcoding_id, report_real_path, scrobble_enabled) select id, name, user_agent, user_name, client, ip_address, last_seen, max_bit_rate, transcoding_id, report_real_path, scrobble_enabled from player;
```

Comment:
AFAICS, you are setting a username in the `player_dg_tmp.user_id` column, and you have a constraint that it should be a valid `user.id`. Is this working?

Reply:
This is handled by the later GO code (since foreign key constraints are disabled during migrations); basically, copy all the items, and either fix the username -> id, or delete it if the username doesn't exist. 
If preferred, I can convert it to an insert in Go-land (it'll just likely be mode code, mostly just for the columns). I did test the upgrade behavior myself though and see the constraints satisfied after the fact.

Reply:
Ah, I see. If we don't have constraints enabled at this point, why can't we do something like this:
```sql
insert into player_dg_tmp(id, name, user_agent, user_id, client, ip, last_seen, max_bit_rate, transcoding_id, report_real_path, scrobble_enabled)
select id, name, user_agent,
       (select id from user where user_name = player.user_name),
       client, ip_address, last_seen, max_bit_rate, transcoding_id, report_real_path, scrobble_enabled from player;
delete from player_dg_tmp where user_id is null;
```

Reply:
That's probably a better idea. Is it guaranteed that all names exist? If not, I can still post-process, and just delete ones where the username is empty string

Reply:
If I understood your question, the usernames that do not exist in the `user` table will yield a null `user_id`. Then we delete the records that have it null in the `delete` step.

Reply:
Will that work with the `user_id varchar not null` constraint?

Reply:
That's a good question... I don't think so. Well, you could do something like:
```sql
ifnull((select id from user where user_name = player.user_name), 'UNKNOWN_USERNAME'),
```
And then:
```sql
delete from player_dg_tmp where user_id = 'UNKNOWN_USERNAME';
```

Reply:
I'll give that a shot, thanks! As an aside, is there a way to test migrations (not manually)?

Reply:
I usually test it manually, as they are relevant only once in the project's lifetime and should not be modified any time after being committed (unless they are written in Go and you need to change method signatures and such)